acceptance: add monotonic inserts test

tbg opened the pull request:
This test loosely mirrors the `monotonic` test from
cockroachdb/jepsen, which has seen a regression in #7899.

Unfortunately, at the time of writing, TestMonotonicIncrements
has not failed, after repeated invocations of

`make acceptance TESTS=Monotonic TESTFLAGS='-v -d 3s -count 10 -nodes=5`.

The bug in #7899 has since been discovered and is covered by a high-level
test which sets up exactly the situation Jepsen has tickled, so it's not
worth optimizing this test to reproduce that exact but any more. See #7955.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7941)

<!-- Reviewable:end -->

knz commented on acceptance/monotonic_insert_test.go:
```diff
@@ -0,0 +1,208 @@
+CREATE DATABASE mono;
+CREATE TABLE IF NOT EXISTS mono.mono (val INT, sts STRING, node INT, tb INT);
+INSERT INTO mono.mono VALUES(-1, '0', -1, -1)`); err != nil {
+		t.Fatal(err)
```
These are in 3 different txns in the original test.

knz commented on acceptance/monotonic_insert_test.go:
```diff
@@ -0,0 +1,208 @@
+		fmt.Println(results)
+	}
+
+	const concurrency = 10
```
Make that 5

tschottdorf replied:
Were they also bound to their respective node? I'm sending them out randomly right now.

knz replied:
Yes they were/are

knz commented on acceptance/monotonic_insert_test.go:
```diff
@@ -0,0 +1,208 @@
+		go func(client mtClient) {
+			invoke(client)
+			<-sem
+		}(clients[rand.Intn(c.NumNodes())])
```
I'd suggest NumNodes different goroutines, each running txns in a (local) loop to a separate note. That would be more like the original workload.

knz commented on acceptance/monotonic_insert_test.go:
```diff
@@ -0,0 +1,208 @@
+	const concurrency = 10
+
+	sem := make(chan struct{}, concurrency)
+	timer := time.After(cfg.Duration)
```
You should only need 5-10 txns to repro. Anything more than that is waste of time.

knz commented on acceptance/replication_test.go:
```diff
@@ -38,6 +38,7 @@ func countRangeReplicas(db *client.DB) (int, error) {
 }
 
 func checkRangeReplication(t *testing.T, c cluster.Cluster, d time.Duration) {
+	return // HACK
```
Yes this is closer to the original test.

knz commented:
Also as expected on Gitter ScanMaxIdleTime should probably be left to the default instead of 200ms

knz commented:
cc @RaduBerinde

tschottdorf commented:
@knz I changed this PR to just supply the test (without trying to tickle the anomaly any more). See updated commit message for details. So nothing interesting here except for catching up our acceptance test coverage to our Jepsen fork a little bit - think that's still worth merging, but I wouldn't fight to the bone about it. PTAL.

tbg committed:
acceptance: add monotonic inserts test

This test loosely mirrors the `monotonic` test from
cockroachdb/jepsen, which has seen a regression in #7899.

Unfortunately, at the time of writing, TestMonotonicIncrements
has not failed, after repeated invocations of

`make acceptance TESTS=Monotonic TESTFLAGS='-v -d 3s -count 10 -nodes=5`.

The bug in #7899 has since been discovered and is covered by a high-level
test which sets up exactly the situation Jepsen has tickled, so it's not
worth optimizing this test to reproduce that exact but any more. See #7955.

knz commented:
LGTM, but please @RaduBerinde have a look at this too and let's strive to reduce code duplication with #7979 if possible.

RaduBerinde commented:
LGTM. I don't know all the particular details of the jepsen test though.

As for #7979 - the main point of that PR is adding the infrastructure that allows us to have arbitrary multi-node parallel tests.  The test I added is a natural extension of the existing single-node test (which I'm told was useful to exhibit failures when certain core changes were in flux).  If both this acceptance test and my test happen to catch the same failure, all the better - debugging failures is probably easier when it's all in a single process (vs multiple docker instances) and easily reproducible with `make test`.

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[acceptance/monotonic_insert_test.go, line 110 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7941#-KNINyv6DI8HW4ep5cuK:-KNINyv6DI8HW4ep5cuL:-135515076) ([raw file](https://github.com/cockroachdb/cockroach/blob/93e6c3d8f81c717404ba0a7892643a48e7a5247d/acceptance/monotonic_insert_test.go#L110)):*
> ```Go
> 			attempt++
> 			l("attempt %d", attempt)
> 			if err := tx.QueryRow(`SELECT cluster_logical_timestamp()`).Scan(
> ```

this things are very hard to read (especially multiline stuff that is part of the `if`). Can't we break it up in a few statements?

```
q = tx.QueryRow('..')
err = q.Scan(..)
if err != nil {..}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7941#-:-KNIOVSWk2s3TccqUjXS:1917161240)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[acceptance/monotonic_insert_test.go, line 110 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7941#-KNINyv6DI8HW4ep5cuK:-KNK6CDp2Hm11dpjkwEp:-2100028921) ([raw file](https://github.com/cockroachdb/cockroach/blob/93e6c3d8f81c717404ba0a7892643a48e7a5247d/acceptance/monotonic_insert_test.go#L110)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> this things are very hard to read (especially multiline stuff that is part of the `if`). Can't we break it up in a few statements?
> 
> ```
> q = tx.QueryRow('..')
> err = q.Scan(..)
> if err != nil {..}
> ```
> 
> 
</details>
I'm surprised that you find these hard to read - I find them pleasing, and at least currently they seem to be in line with convention (grep shows all instances of `QueryRow` in our codebase inline all of this similarly). Maybe there's a discussion to be had here, but I'd rather not have it block this - mind PR'ing your desired changes across `git grep QueryRow` so the aestheticians among us can boulder it out?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7941)*
<!-- Sent from Reviewable.io -->
